pkgs/by-name/xd/xdg-desktop-portal-termfilechooser/package.nix

```diff
@@ -0,0 +1,90 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, makeWrapper
+, fetchpatch
+, meson
+, ninja
+, pkg-config
+, wayland-protocols
+, wayland-scanner
+, inih
+, libdrm
+, mesa
+, scdoc
+, systemd
+, wayland
+,
+}:
+
+stdenv.mkDerivation {
+  pname = "xdg-desktop-portal-termfilechooser";
+  version = "0-unstable-2021-07-14";
+
+  src = fetchFromGitHub {
+    owner = "GermainZ";
+    repo = "xdg-desktop-portal-termfilechooser";
+    rev = "71dc7ab06751e51de392b9a7af2b50018e40e062";
+    hash = "sha256-645hoLhQNncqfLKcYCgWLbSrTRUNELh6EAdgUVq3ypM=";
+  };
+
+  strictDeps = true;
+
+  depsBuildBuild = [ pkg-config ];
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+    scdoc
+    wayland-scanner
+    makeWrapper
+  ];
+
+  buildInputs = [
+    inih
+    libdrm
+    mesa
+    systemd
+    wayland
+    wayland-protocols
+  ];
+
+  mesonFlags = [
+    (lib.mesonOption "sd-bus-provider" "libsystemd")
+    (lib.mesonOption "sysconfdir" "/etc")
+  ];
+
+  patches = [
+    # Allow using the default version, by removing hard-coded paths in the ranger wrapper
+    # https://github.com/GermainZ/xdg-desktop-portal-termfilechooser/pull/9
+    (fetchpatch {
+      name = "dont_hardcode_kitty.patch";
+      url = "https://github.com/GermainZ/xdg-desktop-portal-termfilechooser/commit/af0ae142f54dbf9b480b5c886bfb1319cd6ff25b.patch";
+      hash = "sha256-AgLetBd9sP3G3OPpweVA0Qp9njfPVFXJi4m/rZsSxJc=";
+    })
+    (fetchpatch {
+      name = "dont_hardcode_ranger.patch";
+      url = "https://github.com/GermainZ/xdg-desktop-portal-termfilechooser/commit/4e0db3d4c1639582420847393884ca6bb990cfe5.patch";
+      hash = "sha256-jMNVcwm6zEdOvaTRbw5bfaJ2sLuxfuk7+6DlB1HOkYo=";
+    })
+  ];
+
+  postPatch = ''
+    # Allow using ranger out of the box without any configuration.
+    substituteInPlace src/core/config.c \
+      --replace-fail '"/usr/share/xdg-desktop-portal-termfilechooser/ranger-wrapper.sh"' '"${placeholder "out"}/share/xdg-desktop-portal-termfilechooser/ranger-wrapper.sh"'
+
```

Comment:
Also might want to do something about https://github.com/GermainZ/xdg-desktop-portal-termfilechooser/issues/4 then.

Reply:
Rather than using the main repo and adding 4+ patches I think it would be more suitable to use [an active fork of the repo](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser). Owner of the fork is not a nix user but I have been using his fork for the last couple of months. I also don't mind maintaining a nix specific fork myself if his fork is not suitable for some reason.

Reply:
I'm don't think that this fork, would be able to reduce the needed patches, as it did not include fixes for any of them (listing the patches in the order of the `package.nix`):
- Kitty [is](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser/blob/a3736ca9b3e9c5b19afabfc99f81fbde70a7e066/contrib/ranger-wrapper.sh#L28) still [hardcoded](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser/blob/a3736ca9b3e9c5b19afabfc99f81fbde70a7e066/contrib/fzf-wrapper.sh#L28) (actually more so than in the original repo, as one of the `$TERMCMD` defaults was removed)
-  Ranger is also still [hardcoded](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser/blob/a3736ca9b3e9c5b19afabfc99f81fbde70a7e066/contrib/ranger-wrapper.sh#L27)
- The default ranger wrapper script's path is also still [hardcoded](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser/blob/a3736ca9b3e9c5b19afabfc99f81fbde70a7e066/src/core/config.c#L11)
- And the char to int conversion was [not fixed](https://github.com/boydaihungst/xdg-desktop-portal-termfilechooser/blob/a3736ca9b3e9c5b19afabfc99f81fbde70a7e066/src/filechooser/filechooser.c#L52)

Additionally, issues seem to be disabled in this repo, so notifying them about these things would be difficult.

Reply:
I still believe the fork is useful, as it provides more wrappers, some of them not hardcoded

Reply:
I've started working on an alternative, [termfilepickers](https://github.com/Guekka/xdg-desktop-portal-termfilepickers), written in Rust

It's still very WIP but seems to work

Reply:
I've had a (superficial) look at it and I think that this would be better suited for inclusion in nixpkgs. I could change this pr to adapt it to your portal or close it and let you open a new one, depending on your motivation.

Reply:
@soispha I have written a module. Example usage:
```nix
  imports = [inputs.xdp-termfilepickers.nixosModules.default];

  services.xdg-desktop-portal-termfilepickers = let
    termfilepickers = inputs.xdp-termfilepickers.packages.${pkgs.system}.default;
  in {
    enable = true;
    package = termfilepickers;
    desktopEnvironments = ["hyprland"];
    config = {
      save_file_script_path = "${termfilepickers}/share/wrappers/yazi-save-file.nu";
      open_file_script_path = "${termfilepickers}/share/wrappers/yazi-open-file.nu";
      save_files_script_path = "${termfilepickers}/share/wrappers/yazi-save-file.nu";
      terminal_command = lib.getExe pkgs.kitty;
    };
  };
```

I think this project needs some beta testing before inclusion

For further discussion, please use the discussion tab on the project repo to avoid spamming watchers